### ISSUE TITLE
Any plans to support query load balancing?

### ISSUE THREAD

**Issue by zam6ak** [yandex/odyssey#241]
Hi

Are there any plans for Odyssey to support load balancing of read only queries? 
(for context see [PgPool-II](https://www.pgpool.net/docs/latest/en/html/runtime-config-load-balancing.html)

This would be super useful in cases where there is a streaming replication setup with several replicas in read intensive workloads.
Of course, the trade-off being that query would have to be parsed and analyzed by Odyssey...

Thanks
Zam

**Comment by x4m** [comment]
Hi!
I think the most correct way to do that is to use target_session_attrs in the connection string. We do not need to implement anything for this in Odyssey.
What we have to do though is to implement disconnection of lagging replicas.

**Comment by zam6ak** [comment]
@x4m I am not sure `target_session_attrs` accomplishes same thing as load balancing. 
I would be wrong, if there is an example how to achieve this with that attribute, could you share it please.

 If we consider following  connection string:
`postgresql://host1:123,host2:456/somedb?target_session_attrs=any&application_name=myapp`

Wouldn't application itself need to have knowledge which "type" of query is being sent in order to properly set target_session_attrs? A lot of apps (including ours) currently have static connection string and would have to change before it can send `any` vs `read-write`...

Also [as per docs](https://www.postgresql.org/docs/current/libpq-connect.html),
> It is possible to specify multiple hosts to connect to, so that **they are tried in the given order**.
> ...
> When multiple hosts are specified, or when a single host name is translated to multiple addresses, all the hosts and addresses will be tried in order, **until one succeeds**.

if there are multiple replicas, all connections will go to 1st one specified as long as it able to accept request.
there would be no "balancing"

**Comment by x4m** [comment]
Yes, but you can change order in different service backends. As discussed in the Slack channel. Probably, this is not very good solution. We want something to balance the stream of known-to-be-read-only queries between replicas. This totally makes sense.

**Comment by zam6ak** [comment]
Just to recap, so far, there seem to be two ideas on how to approach this.

One approach, as @x4m indicated is to enable Odyssey to be able to leverage `target_session_attrs`.
This would require that applications connecting have the knowledge of which query is "read only" and to also potentially "rotate" read only replica hosts in connection strings to avoid 1st replica specified getting all the queries (see prev comment about what docs specify)
 
Another approach is to have Odyssey parse the query and determine if it is "read only" and send it to replicas.
And its not only query parsing - proxy would have to potentially query PG catalog to determine which functions (used in queries) are "read only" safe...This is more complicated, and would add overhead to proxy for sure, with advantage that any application connecting via Odyssey could leverage load balancing without code changes...

**Comment by niltooth** [comment]
Is this still open for discussion?

**Comment by rkhapov** [comment]
hi @niltooth and others! actually, there is balancing support in odyssey https://pg-odyssey.tech/features/balancing.html
